Heading block: move alignment controls to toolbar

draganescu opened the pull request:
## Description
Closes #16648

## How has this been tested?
Tested locally.

## Screenshots <!-- if applicable -->
<img width="677" alt="Screenshot 2019-07-19 at 17 56 01" src="https://user-images.githubusercontent.com/107534/61544613-82caa380-aa4e-11e9-8898-cb0f4f3deb26.png">


## Types of changes
Added an alignment toolbar component to the heading block's toolbar.
I think we should also remove them from the inspector, right @mapk ?

youknowriad commented on packages/block-library/src/heading/heading-toolbar.js:
```diff
@@ -9,6 +9,7 @@ import { range } from 'lodash';
 import { Toolbar } from '@wordpress/components';
+import { AlignmentToolbar } from '@wordpress/block-editor';
 
 class HeadingToolbar extends Component {
```
I think that component was just meant to render the heading size contro. Can we just render the alignment toolbar in the `edit.js` file directly instead?

gziolo replied:
Yes, I agree that it should be handled in `edit.js`. `HeadingToolbar` should be called `HeadingSizeToolbar` and handle only that part of the interactions.

draganescu replied:
I moved the align toolbar to the proper location :)

paaljoachim commented:
This issue would then also be associated with moving alignment options of the heading to the toolbar.

Try: Always collapse block alignments.
https://github.com/WordPress/gutenberg/pull/16557

mapk commented:
> I think we should also remove them from the inspector, right @mapk ?

Yep, let's do that. Right now I'm seeing them doubled up. Let's just remove them all now that they are included in the toolbar.

![Screen Shot 2019-07-24 at 4 09 52 PM](https://user-images.githubusercontent.com/617986/61834631-89ca2b00-ae2d-11e9-98be-00c18d79677f.png)

gziolo commented:
> Yep, let's do that. Right now I'm seeing them doubled up. Let's just remove them all now that they are included in the toolbar.

Yes, they should be removed from the sidebar with the proposed change.

We should also make sure that this PR takes into account changes proposed in #16557. I guess it would make sense to land the latter first and then apply it here.

youknowriad commented on packages/block-library/src/heading/edit.js:
```diff
@@ -62,19 +61,11 @@ function HeadingEdit( {
 			<InspectorControls>
-				<PanelBody title={ __( 'Heading Settings' ) }>
-					<p>{ __( 'Level' ) }</p>
-					<HeadingToolbar minLevel={ 1 } maxLevel={ 7 } selectedLevel={ level } onChange={ ( newLevel ) => setAttributes( { level: newLevel } ) } />
```
I think we shouldn't remove the `HeadingToolbar`  from the sidebar because it allows H1 while the block toolbar's one doesn't. (That's done on purpose to deemphasize that heading level.

gziolo replied:
There is also related PR #14889 which adds heading hierarchy checker which will require having those levels in the sidebar to make it easier to fix the issues reported.

draganescu replied:
Oh I see, I'll put them back, wasn't aware of all this, just figured they are duplicated controls. However even with all the explanation, it still feels duplicate.

gziolo commented on packages/block-library/src/heading/heading-toolbar.js:
```diff
@@ -25,7 +25,9 @@ class HeadingToolbar extends Component {
 		const { minLevel, maxLevel, selectedLevel, onChange } = this.props;
 		return (
-			<Toolbar controls={ range( minLevel, maxLevel ).map( ( index ) => this.createLevelControl( index, selectedLevel, onChange ) ) } />
+			<>
```
There is no longer need to use `Fragment` (`<>`).

draganescu replied:
thank you missed that!

youknowriad commented:
This is looking good overall. A few things about consistency and collapsing.

  - The text alignment toolbar is not collapsed in both the paragraph and heading blocks at the root level
 - The text alignment toolbar is collapsed in both the paragraph and heading blocks inside the columns block.
 - The heading level toolbar is not collapsed no matter where we use the blocks.

**Questioins**

- Should we just collapse all these toolbars in all situations especially since we're adding a lot of tools to a lot of blocks?
- If not, should we align the "heading level" toolbar behavior on the text alignment toolbar as well: collapsed inside columns and keep uncollapsed at the root level?

Thoughts @mapk @mtias @jasmussen @kjellr

gziolo commented:
@youknowriad, see this comment under PR I lined earlier today: https://github.com/WordPress/gutenberg/pull/16557#issuecomment-515053335

I think this work started by @jasmussen together with my updates addressed most of the issues you listed.

mapk commented:
> Should we just collapse all these toolbars in all situations especially since we're adding a lot of tools to a lot of blocks?

Yes. I tested @gziolo's additional commits to that PR and the collapsed toolbar elements didn't bother me. It felt natural to open them up even when there was more than one to interact with.

paaljoachim commented:
When there is more then one tool (in the toolbar) associated with doing a similar thing it becomes natural to have them in a drop down. The drop down shows variation of one specific feature.

The more drop down signals an area that has a mix of tools not so often used, but good to have available.

paaljoachim commented:
Btw we should also have the h tags in a drop down in the toolbar.

draganescu commented:
I have removed the align tools from the inspector, added a collapsing item in the toolbar for the heading levels and kept the heading levels always expanded in the inspector.

The problem is that the heading collapsed icon is very confusing: 

<img width="303" alt="Screenshot 2019-08-02 at 21 07 53" src="https://user-images.githubusercontent.com/107534/62391824-88f07200-b56e-11e9-942a-22078dec172d.png">

I am unsure what to use instead as the "H" icon is already there in the block switcher, @mapk any suggestions :D

paaljoachim commented:
I am looking at what it might look like. I am sure there is a better solution.

![h-tag-drop-down](https://user-images.githubusercontent.com/5323259/62395507-c48c3b80-b570-11e9-8bba-654a0ac0bc8f.jpg)

As having two H drop down next to each other can be somewhat confusing.

drw158 commented:
@paaljoachim I like your solution because it shows which level heading is selected. An additional enhancement would be to tweak the H icon on the left to something like this maybe (imagine an H instead the letters shown):
1
![Screen Shot 2019-08-02 at 3 52 13 PM](https://user-images.githubusercontent.com/618551/62398206-a78a4500-b53d-11e9-97e8-9673d40fb54d.png)
2
![Screen Shot 2019-08-02 at 3 51 24 PM](https://user-images.githubusercontent.com/618551/62398207-a78a4500-b53d-11e9-9917-6799ba598690.png)
3
![Screen Shot 2019-08-02 at 3 49 50 PM](https://user-images.githubusercontent.com/618551/62398208-a78a4500-b53d-11e9-83cf-53bc2132a738.png)
4
![Screen Shot 2019-08-02 at 3 49 11 PM](https://user-images.githubusercontent.com/618551/62398209-a78a4500-b53d-11e9-8bd8-1e7b29452382.png)
5
![Screen Shot 2019-08-02 at 3 48 44 PM](https://user-images.githubusercontent.com/618551/62398210-a822db80-b53d-11e9-95c2-50493b24e769.png)
6
![Screen Shot 2019-08-02 at 3 48 18 PM](https://user-images.githubusercontent.com/618551/62398211-a822db80-b53d-11e9-804f-c829edb1f2cf.png)

paaljoachim commented:
Here is the earlier T that is still in use in the Core version of Gutenberg. With a Heading drop down next to it.

![Earlier-version-h-tag-drop-down](https://user-images.githubusercontent.com/5323259/62403738-b4d01f80-b58f-11e9-8cc1-b5a2ff48d3ab.jpg)

<img width="373" alt="Screen Shot 2019-08-03 at 01 39 12" src="https://user-images.githubusercontent.com/5323259/62403740-b863a680-b58f-11e9-967a-7e556bd14e9a.png">


Hey Dave 
@drw158 
Here are the icons you shared added into the wireframe.

![ex1-h-tag-drop-down](https://user-images.githubusercontent.com/5323259/62403796-1d1f0100-b590-11e9-8fce-f3cef128db7c.jpg)

![ex2-h-tag-drop-down](https://user-images.githubusercontent.com/5323259/62403801-1f815b00-b590-11e9-8dea-0f40c057f316.jpg)

![ex3-h-tag-drop-down](https://user-images.githubusercontent.com/5323259/62403804-23ad7880-b590-11e9-8c6c-18c932346640.jpg)

![ex4-h-tag-drop-down](https://user-images.githubusercontent.com/5323259/62403807-26a86900-b590-11e9-8f68-3709d1d76aff.jpg)

![ex5-h-tag-drop-down](https://user-images.githubusercontent.com/5323259/62403810-2a3bf000-b590-11e9-92da-d2abfe39ed90.jpg)

![ex6-h-tag-drop-down](https://user-images.githubusercontent.com/5323259/62403811-2d36e080-b590-11e9-9156-e8a9db02f970.jpg)

drw158 commented:
Thanks for the mockups, that helps.I think the decision to use an `H` instead of `T` is still sound, so we should keep the `H`. We should be able to create a new icon incorporating some of those ideas to make the icon look more like a "group of heading options". 

![Screen Shot 2019-08-06 at 2 22 52 PM](https://user-images.githubusercontent.com/618551/62570841-b16dba00-b855-11e9-8bc9-e45c731d9451.png)

paaljoachim commented:
Here are the sketches added as mockups.

![Gutenberg-Heading-dropdown-h-lines-black-border](https://user-images.githubusercontent.com/5323259/62579371-dee15480-b8a3-11e9-8837-59b161a5aafa.jpg)

![Gutenberg-Heading-dropdown-2-boxes](https://user-images.githubusercontent.com/5323259/62579396-ee609d80-b8a3-11e9-8882-dd4b56b7ea8c.jpg)

![Gutenberg-Heading-dropdown-3-stripes](https://user-images.githubusercontent.com/5323259/62579414-f9b3c900-b8a3-11e9-855a-03c0ec973d52.jpg)

A bonus..:)
![Gutenberg-Heading-dropdown-3-stripes-box](https://user-images.githubusercontent.com/5323259/62579449-14863d80-b8a4-11e9-8152-5010a5b20b87.jpg)


![Gutenberg-Heading-dropdown-h-lines](https://user-images.githubusercontent.com/5323259/62579431-06382180-b8a4-11e9-9fea-5982fcc87527.jpg)

draganescu commented:
I think 5 works best as usually headings clear floats and content is below. Also the other resemble drop cap too much.

richtabor commented:
> I am looking at what it might look like. I am sure not sure if this is a good solution though.
> 
> ![h-tag-drop-down](https://user-images.githubusercontent.com/5323259/62395507-c48c3b80-b570-11e9-8bba-654a0ac0bc8f.jpg)
> 
> Having two H drop downs next to each other can be somewhat confusing.

As long as the H2 icon adapts to switchever is active I’d say this is ok. The hierarchy between the H and following Hx icons works fine here. I vote for keeping the H heading icon as is.

mapk commented:
> I vote for keeping the H heading icon as is.

After looking through all the great mockups from @paaljoachim, I'm more inclined to go this way as well. Nothing else works quite as well as keeping the current icon in addition to the "H2", "H3", etc.

mapk commented:
I just pulled this screenshot from a gif in our documentation here: https://wordpress.org/support/article/wordpress-editor/. Looks like we had the Heading levels in the toolbar at one time before. 😆 

![Screen Shot 2019-08-07 at 6 33 57 AM](https://user-images.githubusercontent.com/617986/62627226-d7d73800-b8dd-11e9-842b-d231302550da.png)

kjellr commented:
Yeah, I think keeping the H icon (at least to start with) seems fine. We can always follow up separately if users find it to be confusing.

richtabor commented:
> I just pulled this screenshot from a gif in our documentation here: https://wordpress.org/support/article/wordpress-editor/. Looks like we had the Heading levels in the toolbar at one time before. 😆
> 
They are present in Gutenberg 6.2.0, just not in a drop-down. :)

paaljoachim commented:
It looks like we can go ahead with the original version with the H in it. Based on feedback above.
![62395507-c48c3b80-b570-11e9-8bba-654a0ac0bc8f](https://user-images.githubusercontent.com/5323259/62686889-edc02800-b9c5-11e9-9c8e-a3555246716f.jpg)

@draganescu

draganescu commented:
@paaljoachim @mapk I am unsure of this: should the headings be always collapsed like the align controls? or collapse only when the view is narrow enough?

richtabor commented:
> @paaljoachim @mapk I am unsure of this: should the headings be always collapsed like the align controls? or collapse only when the view is narrow enough?

I lean towards _always_, to promote the same experience as we have with the alignment controls.

paaljoachim commented:
> @paaljoachim @mapk I am unsure of this: should the headings be always collapsed like the align controls? or collapse only when the view is narrow enough?

As Rich mentions. Headings just like alignments should always be collapsed.

draganescu commented:
These heading buttons are not icons, they are a combination between an H svg icon and a subscript data-attribute  which gets rendered as an after pseudo.

The dropdown menu does not receive nor support rendering the subscript so far and I am working on adding the right way to render that, which should only be some CSS :)

draganescu commented:
I am not 100% happy but here is the current state of things:

![heading-levels](https://user-images.githubusercontent.com/107534/62836758-423a0080-bc6f-11e9-97ec-69af57f6c352.gif)

- there is no bg of the heading dropdown toggle, although no toolbar button behaves like that
- I want to hide the text (heading x and align text etc.) as some CSS side effect made is show up

draganescu commented:
I botched the last commits apparently will fix ASAP :)

draganescu committed:
heading controls update

draganescu committed:
removed useless HoC from heading toolbar since we manually determine collapsing

gziolo commented:
Have you tried to make SVG icon more dynamic using [text](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/text) or something similar for numbers?

mapk commented:
> there is no bg of the heading dropdown toggle, although no toolbar button behaves like that

Good catch, @draganescu! I'd say don't show the background color in the top toolbar level. This way it's consistent with all other toolbar dropdowns. Basically, keep it as you have it.

draganescu commented:
> Have you tried to make SVG icon more dynamic using [text](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/text) or something similar for numbers?

No, but that should be something to check out too. Not sure if in this PR. The H icons are half SVG half subscript :)

draganescu committed:
removed useless function body and some useless export

draganescu commented:
This looks ready to go. @gziolo @mapk what do you think?

draganescu committed:
new snaptshot

mtias commented:
> Thanks for the mockups, that helps.I think the decision to use an H instead of T is still sound, so we should keep the H.

I'm still against using "H" here because, while standard in terms of HTML tags, it perpetuates using English as the canonical source of meaning — the H on its own can be really confusing in other languages for people not familiar with HTML. (I'm totally fine with H being used for the hierarchy level.)

I like some of the ideas proposed by @drw158 for the icon as it gives a bit more context.

Also, the dropdown here seems like a perfect case for showing the actual size rendering of the heading like we used to do on the font-size dropdown. (cc @enriquesanchez)

It would also be a great place to pre-warn if you are choosing an incorrect heading level for the current context.

mapk commented:
> Also, the dropdown here seems like a perfect case for showing the actual size rendering of the heading like we used to do on the font-size dropdown.

This is a great point! All the more reason to get this selectbox right in https://github.com/WordPress/gutenberg/issues/16473#issuecomment-521744718

paaljoachim commented:
It seems this PR which seemed straight forward now has added elements.
1. Not using the H but another icon.
2. Actual size rendering.
These are additional elements that we can get back to later on. Lets get this merged as it is right now as a first iteration. 

Uses the H in the Heading drop down. (H2, H3, H4 etc)
So that we go from this:

![Screen Shot 2019-08-27 at 23 41 24](https://user-images.githubusercontent.com/5323259/63810130-4e3ae900-c924-11e9-98b4-b163241c5820.png)

To something like this:

![Gutenberg-Heading-dropdown-H-Alignment](https://user-images.githubusercontent.com/5323259/63810495-25ffba00-c925-11e9-8f27-d872f110eb25.jpg)

Thanks.
Then we can after this is merged create a new issue for switching the icon and another issue for having actual size rendering of the Heading drop down options.

mapk commented:
I can't get this PR running locally, and so am leaving a proper review to someone who can. 

In light of @paaljoachim's comment above, I'd agree that we should keep this PR focused for now. Moving the heading levels into the dropdown while showing the active heading level in the toolbar's icon looks like a reasonable solution since most everything else in the toolbar is also in a dropdown nowadays.

I'd rather not change the Heading icon back to something else. There was good discussion about this here: https://github.com/WordPress/gutenberg/issues/15340 and seemed to fulfill a11y needs better. I don't believe the "T" was any more cross-language compatible than an "H". And the "H" at least jives with the Heading levels. So let's keep it as is. That being said, I'm totally okay with expanding on the "H" and iconizing it more. But that's another PR.

paaljoachim commented:
The status of this PR is that it is ready for a Code Review, so that we can get it merged.

gziolo commented on packages/components/src/dropdown-menu/index.js:
```diff
@@ -43,6 +43,13 @@ function DropdownMenu( {
 	}
 
+	let activeControl = false;
+	flatMap( controlSets, ( controlSet ) => {
```
It seems like a very specific implementation of the heading dropdown. Can you think of an alternative approach where `DropdownMenu` stays unmodified so we could leave this component as is? Ideally, all those CSS styles are applied directly to `HeadingToolbar` components or alternatively, we modify SVG icons to take the number as param.

draganescu replied:
Maybe, but that would complicate this PR. I tried to follow the same pattern in the interest of pushing the feature forward. We do already use `isActive` in the `DropdownMenu` so I don't feel I've introduced specific Heading related behavior, just some logic to determine the active `DropdownMenu` element.

I will look into implementing the heading icons differently and get rid of the subscript in the heading component but we'll see. I wonder if there is any benefit of still using subscript for these kinds of icons?

draganescu committed:
default collapsed

draganescu committed:
Revert "default collapsed"

This reverts commit d9895aeb9ae28c15311d502de6835b5c4fb3c990.

draganescu committed:
default collapsed

paaljoachim commented:
Hopefully we can get this merged into the next version of Gutenberg. As will have an important effect on the heading UI.

draganescu commented:
i will close this in favor of #17419 which rightfully removes the whole subscript implementation.

gziolo commented:
@draganescu, many thanks for all work on this PR 🙇 

#17419 is now merged.
